### DISCUSSION TITLE
Use constructors as method groups. Add the ability to use constructors as method groups when creating delegates

### DISCUSSION THREAD

**Discussion by jmaine** [dotnet/csharplang#674] [General]
Add the ability to use constructors as method groups when creating delegates.

This:
`Func<object> f = Object.new;`

would be similar to this:
`Func<object> f = Object.StaticMethod;`

as this:
`Func<object> f = () => new object();`

is similar to this:
`Func<object> f = () => Object.StaticMethod();`

A constructor used as a method group when creating a delegate will be act like using a static method group that returns the object that is being constructed when compiling.

**Comment by lachbaer** [comment]
I find this confusing, because it would allow every constructor to be used in this way. I prefer the current way to use an object factory by an explicit static (op-. extension) method named `Create()`.

**Comment by yaakov-h** [comment]
> I find this confusing, because it would allow every constructor to be used in this way.

Isn't that the entire purpose of this proposal/request?

**Comment by GeirGrusom** [comment]
Constructors cannot be bound to a delegate.

edit: to clarify, there is a CLR restriction in place.

**Comment by HaloFour** [comment]
This could be accomplished via syntax candy, though:

```cs
Func<MyClass> func = MyClass::new; // using Java method reference syntax
// compiled into
Func<MyClass> func = () => new MyClass();
```

**Comment by Joe4evr** [comment]
> to clarify, there is a CLR restriction in place.

That sucks, but I recall @CyrusNajmabadi saying in another thread that he *would* like to see this (even suggested `ClassName.new` as the syntax just like this proposal since it's illegal to have a class member named `new`), so there's a possible Champion in the LDM team, at least.

And really, even if it was just syntax sugar as HaloFour suggests, it'd be pretty great.

**Comment by CyrusNajmabadi** [comment]
I would be fine with this *if* there is a clear thing this translates to in the CLR.  And, specifically, we'd need a solution that could ensure the following would work properly:

```c#
x.SomeEvent += SomeType.new;
...
x.SomeEvent -= SomeType.new;
```

**Reply by jnm2** [reply]
I dont believe a non-void-returning delegate makes sense to hook up to a multicast thing like an event.

**Comment by lachbaer** [comment]
> `x.SomeEvent += SomeType.new;`

Where should the result of 'SomeType.new', namely the newly created object, go?

**Comment by jnm2** [comment]
@lachbaer The result is returned from the invocation of the delegate which is backing the event. With multicast delegates, that's the value of either the last or the first invocation (I forget).

**Comment by bondsbw** [comment]
I'll also propose this syntax, for the sake of familiarity:
```c#
Func<MyClass> f = new MyClass;
```

**Comment by GeirGrusom** [comment]
@HaloFour how would you avoid the cross assembly issues?

**Comment by yaakov-h** [comment]
@GeirGrusom What cross-assembly issues?

**Comment by GeirGrusom** [comment]
```csharp
new Func<string, string>(String.Copy) == new Func<string, string>(String.Copy)
```
This evaluates to true today. Imagine if the right hand was done in a different assembly (and it was a constructor instead). It would suddenly evaluate to `false`. And as @CyrusNajmabadi ~~mentioned~~ implied you wouldn't be able to unsubscribe to an event from across assemblies. It would introduce a very sneaky quirk. Or bug as I like to call it.

**Comment by HaloFour** [comment]
@GeirGrusom 

True, the wrapping with a lambda would introduce that problem.  But that's the only way that this feature can be introduced with only language changes.  Constructors are instance `void` methods and you're not permitted (well, not *supposed to*) invoke them directly.  To treat them like a static factory you'd need something to wrap the constructor.

While the event subscription situation does exist I'd wager that it would be exceptionally rare.  Constructors don't fit in the normal convention for defining events, and events that return values, while technically legal in C# (not in VB.NET), are non-deterministic in multiple subscription scenarios.

**Comment by GeirGrusom** [comment]
> To treat them like a static factory you'd need something to wrap the constructor.

Or CLR support.

**Comment by lostmsu** [comment]
I like @bondsbw's syntax more, if possible.
It is incorrect, that you can't define a member called `new`. Just do `@new`, and the bytecode will have it named `new`.

**Comment by ghost** [comment]
`Func<MyClass> f = new MyClass;`
This could act exactly the same as static functions work, and I can't think of any scenarios where this is currently valid syntax.

**Comment by jamesharling** [comment]
I was about to open an issue to suggest this feature. I would also favour @bondsbw's syntax, it is in keeping with both current constructor syntax and method group syntax.

I was surprised to find constructors not supported as method groups, something like the following feels like it should be supported C#:

```csharp
public class StringWrapper
{
    public string Value { get; }

    public StringWrapper(string value) => Value = value;
}

var strings = new string[] { "foo", "bar" };

var wrappers = strings.Select(new StringWrapper);

**Comment by MarcSigrist** [comment]
I would welcome this feature in C#, because

- I have used it a lot in F#. 
- Intuitively, it feels like one of the last "missing pieces", considering the strong trend to remove needless syntactic overhead from C#, such as e.g. with index initializers (C# 6.0), expression-bodied constructors (C# 7.0), using declarations (C# 8.0), "Extension Everything" (proposed), ...

From what I have seen so far, I would prefer @bondsw's syntax.

**Comment by TheBrambleShark** [comment]
I think I would prefer `Type::new` ~~or `Type::ctor`~~. If it's just syntax candy, it should fit really well with things like `Select(Foo::new)` -> `Select(it => new Foo(it));`

I think the only outstanding question would be overload resultion in case of multiple constructors, but that should be something that can be handled just like `Select(Foo.SomeStaticMethod)`

`Select(new Foo)` doesn't feel quite right, so I'd vote against that syntax.

Edit: @333fred points out that `Foo:ctor` could cause conflicts if you have an extern-aliased assembly named `Foo` and a type `ctor` in that assembly.

**Comment by LWChris** [comment]
I would prefer `Type.new` as syntax. C# has a syntax for "all methods named `Y` on the type `X`", and that is `X.Y`. I don't see any reason to invent a new syntax, just because it's the method group of constructors. The only "invention" required is a name for that method group, since constructors don't really have a name. And `Type.new` feels like a perfect fit for that name. The only other candidate C# syntax suggests is `Type.Type`, but that is longer for almost every class name and literally redundant.

`Type.new` cannot create a conflict, because this feature is not about adding an actual method called `new` which _could_  conflict with the byte code method name for `@new`; it is about syntactic sugar to denote a method group of constructors. And for C# syntactic sugar, the only consideration is: can there - today in C# - be a case, where `Type.new` was legal before? And to that, the answer is "no", because any existing C# code must be `Type.@new` which does not conflict with the proposed syntax `Type.new`.

**Comment by HaloFour** [comment]
If runtime support is required for this to be considered that feels like it makes it dead on arrival.  Is that support required, though?  F# already allows using primary constructors like function calls by wrapping the constructor call.

**Reply by BlinD-HuNTeR** [reply]
I was thinking about this feature, and got some crazy ideas...

First of all, the original answer for this issue is "not possible". Which is, kinda, subject to interpretation. In software, nothing is truly impossible, and of course many people just suggested that "compiler rewrites constructor method group into a lambda". Which would work, of course.

The problem is that, this would violate an existing assumption in the language: when you use a "**method group**" to create a delegate, instead of a lambda, you are **guaranteed** that `Delegate.Method` returns the exact method you passed in source code. And this assumption is very important for event handlers, because `Delegate.Equals` compares both the `Target` and the `Method` when removing an event handler.

And for the case of constructors, you can't really have a delegate pointing directly to a constructor. That's because a constructor is actually a void-returning method, that doesn't return the new object, but instead **takes** the new object as its `this` parameter. So, in order to have a delegate that creates objects, the compiler would need to generate a wrapper method, and that method is what would be returned by `Delegate.Method`.

Now, an argument could be made that all of this doesn't really matter, because no one really uses constructors for event handlers, so, the people who want this feature don't care if `Delegate.Method` returns something else. So in theory, this would just be a spec change, not a runtime change (I don't even know if the spec even says anything about `Delegate.Method`)

And maybe, to avoid misuse, the language could even forbid that constructor method groups be used in event accessors (code like `someObject.someEvent += MyType.new` would then give a compiler error saying "constructors can't be used as event handlers"). This would force the user to assign the delegate to a variable before, and then, that variable could be safely used to add and remove an event handler.

But then I got another idea: what if `Delegate.Method` could actually return the true constructor, even though the delegate actually points to a wrapper method instead?

Then I found this hidden `MulticastDelegate` constructor in the corlib:
```Cs
public abstract class MulticastDelegate : Delegate
{
    private void CtorCollectibleClosedStatic(Object target, IntPtr methodPtr, IntPtr gchandle)
    {
        this._target = target;
        this._methodPtr = methodPtr;
        this._methodBase = System.Runtime.InteropServices.GCHandle.InternalGet(gchandle);
    }
}
```
So, if the compiler could invoke this constructor, it could make the delegate point to an implicit static method, while at the same time, return the true constructor for `Delegate.Method`.

Unfortunately, though, this constructor is private, so it would have to be made public (actually it's not even a constructor, but rather an initializer invoked by the runtime), so, this is the part that would require runtime support, but as I said, I believe most people don't really care about this.

**Reply by jnm2** [reply]
I dont believe a non-void-returning delegate makes sense to hook up to a multicast thing like an event.

**Reply by LWChris** [reply]
Yes, it doesn't make sense, but it is allowed. And even if we concluded that "Events wouldn't work but multicast delegates don't play nicely with non-void anyway", then this is a thing:

```csharp
public class FactoryCollection<T>
{
    private readonly List<Func<T>> _factories = [];
    public void Add(Func<T> factory) => _factories.Add(factory);
    public bool Remove(Func<T> factory) => _factories.Remove(factory);
    public IEnumerable<T> Execute() => _factories.Select(f => f());
}
```

This would work with method group delegates but not with "method group lambdas".
Therefore, as much as it pains me to admit:

If I have

```csharp
public class Foo
{
    public Foo() { }
    public Foo(string s) { }
    public Foo(int i) { }

    public static Foo Create() => new();
    public static Foo Create(string s) => new(s);
    public static Foo Create(int i) => new(i);
}
```

then `Foo.Create` must be 100% exchangable with `Foo.new` for this to feel right. And a lambda just doesn't meet that requirement.

**Reply by BlinD-HuNTeR** [reply]
>This would work with method group delegates but not with "method group lambdas".

If the compiler rewrites the constructor method group into a lambda, then it can reuse the same lambda instance for every usage of a given constructor. This way, `Add` and `Remove` would work just fine **as long as you are adding and removing from the same assembly**.

If you still want your "FactoryCollection" to support removal of a delegate, from a different assembly that added it, then this is impossible to do without runtime support.

Anyway, you have to pick your poison.
